### DISCUSSION TITLE
안녕하세요. 문의드립니다.

### DISCUSSION THREAD

**Discussion by tpag20** [elbakramer/koapy#5] [Q&A]
새해복 많이 받으시기 바랍니다.
너무 훌륭한 프로그램을 공개해주셔서 대단히 감사합니다. 파이썬으로 프로그래밍 입문하여 현재 기본 문법 정도만 익힌 왕초보 입니다. 할 줄 아는건 웹스크래핑 정도입니다. 

현재 koapy를 이용해서 디버깅을 하면서 공부하며 제작을 하고있는데 gRPC 환경이라서 디버깅이 매우 어려운 상황입니다.(이걸 알아내는데도 한참 걸렸습니다ㅠ) 중단점에서 프로그램이 일시정지 되어도 세션?이 닫혀서 데이터를 불러올수가 없는 것 같은데 

> 1.  혹시 디버깅하면서도 데이터를 가져올 수 있는 방법이 없을까요? 예를들면 디버깅으로 중단점에서 멈춰놓고 주식 종목을 가져오는것 처럼요.

> 2.  한가지 더. 프로그램을 실행할때마다 로그인창이 계속 뜨는데 이미 키움 open api에 로그인 되어있을때는 안뜨도록 할 수 없을까요? 

너무 하찮은 질문 드려서 죄송합니다. 하지만 이것 때문에 앞으로 나아가질 못하고 있습니다. 저에게 너무 중요합니다.  문서 보는 방법도 모르겠고 많이 힘드네요. 오늘 koapy설치하고 주식종목 불러와서 내림차순으로 정렬시키는데 7시간 걸렸답니다. ㅠㅠ 아무쪼록 위의 질문에 답변을 주시면 정말 감사하겠습니다. 약소하지만 커피 한잔 값을 후원합니다. 즐거운 저녁되세요

**Comment by tpag20** [comment]
이럴수가 해외결제 카드가 없어서 후원이 불가능합니다 ㅠㅠ 카톡아이디나 다른 방법을 알려주세요. 제가 원하는대로 프로그램 제작을 다 마치고 나면 또 후원을 하도록 하겠습니다.

**Comment by elbakramer** [comment]
안녕하세요.

#### 1. 혹시 디버깅하면서도 데이터를 가져올 수 있는 방법이 없을까요? 예를들면 디버깅으로 중단점에서 멈춰놓고 주식 종목을 가져오는것 처럼요.

제가 말씀주신 것 처럼 중단점을 설정해서 하는식의 디버깅 기능을 사용해본적이 잘 없어서 관련해서는 조언드릴 수 있는게 별로 없네요;;
저는 그냥 무식하게, 확인하고 싶은게 있는 부분에 print 같은걸 넣어가면서 계속 수차례 돌려보는 식으로 하고 있어서요.

일반적인 `KiwoomOpenApiContext` 객체를 사용하는 방식으로 하고 계신다면 말씀주신 방식으로 디버깅 하기가 좀 어려울거 같긴 합니다.
해당 객체를 사용하면 무조건 gRPC 의 서버/클라이언트 구조를 사용하도록 되어있어서 디버깅을 한다고 해도 아마도 서버/클라이언트 각각 해서 두벌로 해야 할거에요.

추측하기로는 맥락상 IDE 같은 환경에서 제공하는 실행버튼 같은거로 스크립트를 실행을 하시거나 아니면 스크립트 하나를 가지고 돌리시는 듯 한데요. 이런 경우에는 `KiwoomOpenApiContext` 내부에서 subprocess 형태로 서버를 자기가 열고 자기가 접속하는 형태를 취하기 때문에 더더욱 서버쪽 디버깅은 어렵지 않을까 싶습니다. (클라이언트 입장에서 중단점을 걸더라도 서버는 별도 프로세스에서 알아서 돌아가지 않을지..)

아래 2번에 답변드리는 내용을 참고하서서 서버랑 클라이언트를 각각 따로 띄우고 따로 디버깅하는 형태로 한번 시도해보세요.

그리고 딱히 서버/클라이언트 구조문제가 아니더라도 현재 이벤트처리 같은 경우 비동기/쓰레드 방식으로 처리가 될텐데요. 여기에다 중단점을 걸어서 디버깅을 하는식이 되는건지도 잘 모르겠네요;;

#### 2. 한가지 더. 프로그램을 실행할때마다 로그인창이 계속 뜨는데 이미 키움 open api에 로그인 되어있을때는 안뜨도록 할 수 없을까요?

키움 Open API 에서는 프로그램이 실행하지 않더라도 항상 떠있는 그런게 따로 없어서요.
키움 Open API 가 탑재된 프로그램이 죽는 순간 다른 키움 관련도 다 죽어버리는 형태라 사실 이미 로그인 되어있다는 개념이 딱히 없습니다.

이런게 제가 gRPC 서버/클라이언트 형태를 생각한 이유중에 하나인데요.
핵심기능만 구현된 서버를 따로 띄워놓고 클라이언트가 접속하는 형태가 되면, 클라이언트가 죽더라도 서버쪽은 살아있어서 로그인이 유지되는 형태입니다. 서버쪽에서 뭔가 에러가 나서 죽지 않는 이상 로그인은 유지될겁니다.

따라서 이거는 아래 문서를 참고하신뒤에, CLI 를 활용하셔서 별도 서버를 띄워놓고 사용하는 방식을 추천드립니다.
https://koapy.readthedocs.io/en/latest/usage.html#grpc-server

혹시나 CLI 가 익숙하지 않으시면 아래 스크립트를 실행하셔도 됩니다.
https://github.com/elbakramer/koapy/blob/master/koapy/pyside2/tools/start_tray_application.py
